tests/feature/Issues/Issue1427Test.php

```diff
@@ -0,0 +1,93 @@
+      - `.16` must pass the rules
+        - Each item in `.groups` must be valid
+          - `.1` must be an integer value
+        - Each item in `.permissions` must be valid
+          - `.perm3` must be a boolean value
+      - `.18` must pass all the rules
+        - `.18` must be an array value
+        - `.groups` must be present
+        - `.permissions` must be present
+      - `.24` must pass the rules
+        - `.groups` must be present
+        - `.permissions` must be iterable
+    FULL_MESSAGE,
+    [
+        'messages' => ['each' => 'Each item in `[16: ["groups": [1, "A", 3, 4, 5], "permissions": ["perm1": true, "perm2": false, "perm3": "boom!"]], 18: false, ... ]` must be valid'],
+        'children' => [
+            16 => [
+                'messages' => ['allOf' => '`["groups": [1, "A", 3, 4, 5], "permissions": ["perm1": true, "perm2": false, "perm3": "boom!"]]` must pass the rules'],
+                'children' => [
+                    'groups' => [
+                        'messages' => ['each' => 'Each item in `[1, "A", 3, 4, 5]` must be valid'],
+                        'children' => [
+                            1 => [
+                                'messages' => ['intVal' => '"A" must be an integer value'],
+                            ],
+                        ],
+                    ],
+                    'permissions' => [
+                        'messages' => ['each' => 'Each item in `["perm1": true, "perm2": false, "perm3": "boom!"]` must be valid'],
+                        'children' => [
+                            'perm3' => [
+                                'messages' => ['boolVal' => '"boom!" must be a boolean value'],
+                            ],
+                        ],
+                    ],
+                ],
+            ],
+            18 => [
+                'messages' => ['allOf' => '`false` must pass all the rules'],
+                'details' => ['arrayVal' => '`false` must be an array value'],
+                'children' => [
+                    'groups' => [
+                        'messages' => ['keyExists' => '`false` must be present'],
+                    ],
+                    'permissions' => [
+                        'messages' => ['keyExists' => '`false` must be present'],
+                    ],
+                ],
+            ],
+            24 => [
+                'messages' => ['allOf' => '`["permissions": false]` must pass the rules'],
+                'children' => [
+                    'groups' => [
+                        'messages' => ['keyExists' => '`["permissions": false]` must be present'],
+                    ],
+                    'permissions' => [
+                        'messages' => ['each' => '`false` must be iterable'],
+                    ],
+                ],
+            ],
+        ],
+    ]
```

Comment:
I gotta say, we can added a whole bunch of other things here, I'm just not sure how they can be useful. Each element could contain the keys:

* `messages`: all the messages that fails
*`details`: more details about the messages that failed
* `children`: all the children (which in turn could have all the keys I'm mentioning here
* `input`: the input that was proceeded by the rule
* `path`: the path of the rule that failed

But when I start to really think about this, I think this is not even an exception anymore, it should be an object that contains all this information and handles all of them properly. It could very well be the case.

Maybe I'm just thinking of providing this array because of how the library worked before, but I could just drop the whole array thing, and return a proper object in `Validator::validate()` that would contain all the failed messages and the extra information that is needed -- maybe I'm just conditioned to how the library worked in the past.

Reply:
I can offer some perspective on what I was thinking when I made the original library.

The array thing and a lot of things (like setting properties in runtime objects) were implementation details, mostly private stuff, that could be done in whatever way fits best (I decided to go for the solution that had lass code back then).

To me, the `reportError` (or was it `reportException`?) method was part of the validatable interface that could be extended and the thing I wanted to look nice. Together with the base lib, it offered two standard ways of creating exceptions:

  - ValidationException, which is a simple single message object.
  - AbstractNestedException, which is a more elaborate object, mostly used by aggregate validators, but actually an object of its own.

AbstractNestedException had things like `findMessages` with paths back then, although the workings were slightly different. Since each validator had to had a dedicated exception, I went for an abstract instead of a concrete named object like `ValidationResults`.

The decision to use exceptions also played a role in the design of `Validatable::check`, which I think doesn't exist anymore. It was designed to catch exceptions by type early instead of running the whole thing (the decision whether to fail fast or not). This complicated the internal workings of the project, and it was a compromise I'm not sure we still need to make. It seems our users don't care about fail fast that much. 

So, you're still dealign with a lot of interlocked design decisions I made back then :D. You don't need to follow any of them if they don't make sense anymore. I would just go for whatever fullfills your vision best. That was what I did: I just moved things around until it was cohese.